### ISSUE TITLE
dbt-materialize: error parsing latest version from PyPi

### ISSUE THREAD

**Issue by morsapaes** [MaterializeInc/materialize#11417]
### What version of Materialize are you using?

v0.24.0

### How did you install Materialize?

Docker image

### What was the issue?

Installing (or upgrading to) the latest version of `dbt-materialize` leads to:

`Could not parse version "1.0.3.post1"`

To reproduce :

```bash
pip install dbt-materialize #or pip install --upgrade dbt-materialize==1.0.3.post1

dbt --version
```

I made sure that `dbt-core` and `dbt-postgres` were bumped to `1.0.3`:

```bash
pip list

Package                  Version
------------------- -----------
dbt-core             1.0.3
dbt-extractor        0.4.0
dbt-materialize      1.0.3.post1
dbt-postgres         1.0.3
```

and that I'm on Python 3.9:

```bash
python --version
Python 3.9.10
```

~The same error is thrown for `1.0.3`.~ Versions up to `1.0.3` can be installed with no issues (including previous `.post` versions).

<hr>

It doesn't seem like there were any changes to dbt's handling of SemVer (see [`/core/dbt/semver.py`](https://github.com/dbt-labs/dbt-core/commits/main/core/dbt/semver.py)), so it's not immediately clear to me what's up. There was a [similar issue](https://github.com/dbt-labs/dbt-core/issues/3109) reported last year, but it was related to an invalid semantic version identifier.

This issue has also been reported in the [dbt community Slack](https://getdbt.slack.com/archives/CBSQTAPLG/p1648184495967329).

**Comment by morsapaes** [comment]
In a fresh `venv`, managed to get `1.0.3` installed, but still running into issues with `1.0.3.post1`. Sigh, this might end up just being some Python environment configuration weirdness.

**Comment by benesch** [comment]
The issue seems to have been https://github.com/dbt-labs/dbt-core/commit/f3c7b6bfd158df9cda3077a090f430fb9cd6e147. Basically, `dbt --version` now attempts to parse the dbt version as SemVer before it displays it; previously it just blindly emitted the version. AFAICT this only impacts `dbt --version`. With v1.0.3.post1, `dbt run` works just fine for me; only `dbt --version` complains.

I don't have any good solution for fixing this. I'm not sure there is one. We could potentially move to pre-releases instead of post-releases. Like, our next release could be v1.0.5-pre1 rather than v1.0.4.post1. But pre-releases imply instability, which isn't really at play here.

The best option might be to just queue up changes that we previously would have released as a `post` version, and just wait for the next release of dbt to drop.

**Comment by benesch** [comment]
I'm sure @jtcohen6 would have some useful advice here!

**Comment by jtcohen6** [comment]
Thanks for tagging me in @benesch!

You're right, it looks like we now care to validate that each plugin's version is semantic, for the purposes of being able to say "Update available!" or "Up to date!" Looks like you could fix this by starting to use `1.0.3+post1` (`+` meaning ["build" metadata](https://semver.org/#spec-item-10)) instead.

For my part, I'd encourage you to follow our lead by adjusting this dependency:
https://github.com/MaterializeInc/materialize/blob/106abf2415ace9e469a5fb0cbf4daccee6e99013/misc/dbt-materialize/setup.py#L44
To use a [compatibility operator (`~=`)](https://peps.python.org/pep-0440/#compatible-release) instead:
```python
 # always install latest dbt-postgres==1.0.x + dbt-core==1.0.x
 install_requires=["dbt-postgres~=1.0.0"],
```
And to start releasing totally new patch releases, rather than futzing with pre/post/build metadata specifiers. (You're of course still welcome to put out actual prereleases of patches, e.g. v1.0.5-rc1.)

Starting in v1.0, we're trying to set users's expectations that:
- Minor versions matter, may include some behind-the-scenes changes (non-breaking for end users), and should be coordinated across `dbt-core` + plugins — i.e. you should be using v1.1.x of `dbt-core` together with v1.1.x of `dbt-materialize`
- Patch releases are fixes only, and patch numbers may differ across dbt-core + installed plugins. As long as you're using the latest one, you're good!

**Comment by morsapaes** [comment]
Thanks for jumping in here and in the dbt community [thread](https://getdbt.slack.com/archives/C030A0UF5LM/p1649275178019809), @jtcohen6! Your answers are always so clear and helpful.

@benesch, it sounds like your suggestion to queue up changes would play nicely with the new versioning rules; we could just release patches once there's a meaningful scope of changes and drop the `.post` shenanigans. WDYT? This would also force us to be more intentional each `dbt-materialize` release, which I like!

(CC @ahelium)

**Comment by benesch** [comment]
Yup, sounds good to me! (I think we can still cut a new patch version for every change, if we wanted; sounds like nothing would go wrong if we're on 1.0.29 while dbt-core is still on 1.0.4. But being more intentional about our releases certainly SGTM.)

> Looks like you could fix this by starting to use 1.0.3+post1 (+ meaning ["build" metadata](https://semver.org/#spec-item-10)) instead.

AIUI build metadata doesn't impact the version comparison. So `1.0.3+post1` is the same version as `1.0.3+post2`, as far as SemVer is concerned! Totally moot, though; we'll just let our patch numbers drift. Thanks very much, @jtcohen6, as usual!

**Comment by morsapaes** [comment]
Fixed in #11613. Thanks to everyone involved in the discussion!

We followed @jtcohen6's suggestion of switching to the compatibility operator, and will be cutting patch releases as we see fit.
